### PR TITLE
Implement Precision operator for DateTime/Date/Time

### LINKED ISSUE #160: Support "Precision" operator
See: https://cql.hl7.org/09-b-cqlreference.html#precision

### PR DESCRIPTION
**Pull request by kjmahalingam**
Various fixes to tests and/or logic to address the following test failures:

| **Test** | **Error** |
| --- | --- |
| `CqlArithmeticFunctionsTest.Precision.PrecisionYear` | Precision not implemented |
| `CqlArithmeticFunctionsTest.Precision.PrecisionDateTimeMilliseconds` | Precision not implemented |
| `CqlArithmeticFunctionsTest.Precision.PrecisionTimeMinutes` | Precision not implemented |
| `CqlArithmeticFunctionsTest.Precision.PrecisionTimeMilliseconds` | Precision not implemented |

Does *not* fix `PrecisionDecimal` since this requires additional work with integers and decimals as in #204 

Fixes #160 

Pull requests into cql-execution require the following.
Submitter and reviewer should ✔ when done.
For items that are not-applicable, mark "N/A" and ✔.

[CDS Connect](https://cds.ahrq.gov/cdsconnect) and [Bonnie](https://github.com/projecttacoma/bonnie) are the main users of this repository.
It is strongly recommended to include a person from each of those projects as a reviewer.

**Submitter:**
- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [ ] Code coverage has not gone down and all code touched or added is covered.
- [x] Code passes lint and prettier (hint: use `yarn run test:plus` to run tests, lint, and prettier)
- [x] All dependent libraries are appropriately updated or have a corresponding PR related to this change
- [x] `cql4browsers.js` built with `yarn run build:browserify` if source changed.

**Reviewer:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: src/datatypes/datetime.js [hunk 1]
```diff
@@ -342,6 +342,25 @@ class DateTime {
...
+      precisionValueMap.set(DateTime.Unit.YEAR, 4);
+      precisionValueMap.set(DateTime.Unit.MONTH, 6);
+      precisionValueMap.set(DateTime.Unit.DAY, 8);
+      precisionValueMap.set(DateTime.Unit.HOUR, 10);
+      precisionValueMap.set(DateTime.Unit.MINUTE, 12);
+      precisionValueMap.set(DateTime.Unit.SECOND, 14);
+      precisionValueMap.set(DateTime.Unit.MILLISECOND, 17);
+    }
+    return precisionValueMap.get(this.getPrecision());
+  }
```
Comment (cmoesel) [inline src/datatypes/datetime.js:362]: I think it would be more performant to initialize these maps as constants in the file -- and then just do something like:
```js
const pvMap = this.isTime() ? TIME_PRECISION_VALUE_MAP : DATETIME_PRECISION_VALUE_MAP;
return pvMap.get(this.getPrecision());
```

File: src/datatypes/datetime.js [hunk 2]
```diff
@@ -623,6 +642,14 @@ class Date {
     return result;
   }
 
+  getPrecisionValue() {
+    const precisionValueMap = new Map();
+    precisionValueMap.set(DateTime.Unit.YEAR, 4);
+    precisionValueMap.set(DateTime.Unit.MONTH, 6);
+    precisionValueMap.set(DateTime.Unit.DAY, 8);
+    return precisionValueMap.get(this.getPrecision());
+  }
```
Comment (cmoesel) [inline src/datatypes/datetime.js:651]: Same here -- I'd recommend initializing a constant map, or you could probably use the same map as DateTime uses.

File: src/elm/arithmetic.js [hunk 3]
```diff
@@ -440,7 +440,7 @@ class Successor extends Expression {
   exec(ctx) {
     const arg = this.execArgs(ctx);
     if (arg == null) {
-      null;
+      return null;
```
Comment (cmoesel) [inline src/elm/arithmetic.js:443]: Whoops!  Good catch!

**Review by cmoesel — CHANGES_REQUESTED** [review]
Implementation looks good, but I'd prefer an approach that doesn't create a new Map every time the function is invoked.

File: src/datatypes/datetime.js (hunk 1 above)
Comment (kjmahalingam) [reply src/datatypes/datetime.js:362]: I actually did something like this initially, and then realized that for some reason the `DateTime` namespace was out of scope of where I was defining them at the top of the file. Maybe I can define the maps within the classes themselves, and that should take care of the issue. I'll give it a shot!

File: src/datatypes/datetime.js (hunk 2 above)
Comment (kjmahalingam) [reply src/datatypes/datetime.js:651]: Will do the same here!

File: src/datatypes/datetime.js (hunk 1 above)
Comment (kjmahalingam) [reply src/datatypes/datetime.js]: Turns out the issue was an ordering thing; I put them at the top of the file before `DateTime.Unit` was available. I just put them with the other similar constants and now it's all working.
